### PR TITLE
Biovotion device integration

### PR DESCRIPTION
**Pull request by sboettcher**
Files for the Biovotion VSM device and integration into Main etc.
I replaced the Angel sensor row, since it won't be used anymore as far as I know.

Outstanding issues so far:

- No accelerometer raw data, not currently sent from the device. I am waiting on a response from the Biovotion team on that.
- The UI is not updating after a reconnect of the device (e.g. after going out of range and back). Data is still recieved and sent to the the server though. I'm not really sure if this is a mistake in my code or maybe some other bug.

.aar library files are available on request, just send me a short mail.

### CODE REVIEW

**Comment by blootsvoets** [comment]
This looks very complete indeed, thanks a lot! A thing you might try to get the connection UI working properly is:

when you receive a connection error in `BiovotionDeviceManager`, do a `updateStatus(DISCONNECTED)`. Plus, perhaps, do you need to unbind from the gattService in `close` or when `updateStatus(DISCONNECTED)` is called?

**Comment by sboettcher** [comment]
Unfortunately unbinding the gatt service did not help. As far as I can see, the "problem" is in the update method of the MainActivityUiUpdater, currently line 109: `deviceData[i] = connection.getDeviceData();`.

After a reconnect, `connection` has the correct device status `CONNECTED` (`DeviceStatusListener$Status`), but the device status returned by `getDeviceData()` (`BiovotionDeviceStatus`) is the default status with NaN's and `READY`.

It seems to me like at some point a new `BiovotionDeviceStatus` object is created but the `DeviceServiceConnection` in the MainActivity is not updated?

File: app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+    @Override
+    public void onVsmDeviceConnecting(@NonNull VsmDevice device) {
+        logger.info("Biovotion VSM device connecting.");
+        updateStatus(DeviceStatusListener.Status.CONNECTING);
+    }
+
+    @Override
+    public void onVsmDeviceConnectionError(@NonNull VsmDevice device, VsmConnectionState errorState) {
+        logger.error("Biovotion VSM device connection error.");
```
Comment (blootsvoets) [inline app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java]: could you add a
```java
updateStatus(DeviceStatusListener.Status.DISCONNECTED);
```
here? That way the device manager should shut down this device manager and the MainActivity will start a new one. For more information while debugging, perhaps it's useful to write
```java
logger.error("Biovotion VSM device connection error ({}).", errorState);
```
here?

**Comment by blootsvoets** [comment]
The unbinding is one part, another is that the DeviceService assumes a DeviceManager object is used for exactly one connection. So if it fails, it should update the status to disconnected, so it can restart the connection. I hope that helps in resolving the error...

File: app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java (hunk 1 above)
Comment (sboettcher) [inline app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java]: yes I had already tried that, sorry I forgot to push. The thing is that on a disconnect the error callback isn't called, but the disconnect, which already had the updateStatus().

File: app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+            return;
+
+        logger.info("Biovotion VSM device connected.");
+
+        updateStatus(DeviceStatusListener.Status.CONNECTED);
+
+        vsmDevice = device;
+
+        vsmStreamController = device.streamController();
+        vsmStreamController.addListener(this);
```
Comment (blootsvoets) [inline app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java]: At this point the vsmDevice gets a StreamListener attached. Could it be that this never gets unattached in the Biovotion SDK?

File: app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        }
+    }
+
+
+    public void close() {
+        synchronized (this) {
+            if (this.isClosed) {
+                return;
+            }
+            logger.info("Closing device {}", deviceName);
```
Comment (blootsvoets) [inline app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java]: Is this called?

File: app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+            vsmDevice.setBleService(vsmBleService);
+
+            logger.info("Biovotion VSM initialize BLE service");
+            if (!vsmBleService.initialize(context.getApplicationContext()))
+                logger.error("Biovotion VSM unable to initialize BLE service");
+
+            // Automatically connects to the device upon successful start-up initialization
+            String id = vsmDevice.descriptor().address();
+            logger.info("Biovotion VSM Connecting to {} from activity {}", vsmDevice.descriptor(), this.toString());
+            vsmBleService.connect(id, VsmConstants.BLE_CONN_TIMEOUT_MS);
```
Comment (blootsvoets) [inline app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java]: Do we need to disconnect this as well when the service is disconnected?

**Comment by blootsvoets** [comment]
I'm also not sure where the problem might be... I've added some pointers, but I can't be sure.

File: app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java (hunk 2 above)
Comment (sboettcher) [inline app/src/main/java/org/radarcns/biovotionVSM/BiovotionDeviceManager.java]: Just got it to work! It was actually this and the `vsmDevice` listeners not being removed. Thank you so much for the pointers! Will push fixing commit shortly.

**Pull request merged**